src/transformers/models/dpt/modeling_dpt.py

```diff
@@ -0,0 +1,1127 @@
+                    def custom_forward(*inputs):
+                        return module(*inputs, output_attentions)
+
+                    return custom_forward
+
+                layer_outputs = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(layer_module),
+                    hidden_states,
+                    layer_head_mask,
+                )
+            else:
+                layer_outputs = layer_module(hidden_states, layer_head_mask, output_attentions)
+
+            hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (layer_outputs[1],)
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states, all_self_attentions] if v is not None)
+        return BaseModelOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+        )
+
+
+class DPTReassembleBlocks(nn.Module):
+    """
+
+    ViTPostProcessBlock, process cls_token in ViT backbone output and rearranges the feature vector to feature map.
+
+    Args:
+        in_channels (int): ViT feature channels. Default: 768. out_channels (List): output channels of each stage.
+            Default: [96, 192, 384, 768].
+        readout_type (str): Type of readout operation. Default: 'ignore'. patch_size (int): The patch size. Default:
+        16. init_cfg (dict, optional): Initialization config dict. Default: None.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+
+        self.config = config
+        out_channels = config.post_process_channels
+
+        self.projects = nn.ModuleList(
+            [
+                nn.Conv2d(in_channels=config.hidden_size, out_channels=out_channel, kernel_size=1)
+                for out_channel in out_channels
+            ]
+        )
+
+        self.resize_layers = nn.ModuleList(
+            [
+                nn.ConvTranspose2d(
+                    in_channels=out_channels[0], out_channels=out_channels[0], kernel_size=4, stride=4, padding=0
+                ),
+                nn.ConvTranspose2d(
+                    in_channels=out_channels[1], out_channels=out_channels[1], kernel_size=2, stride=2, padding=0
+                ),
+                nn.Identity(),
+                nn.Conv2d(
+                    in_channels=out_channels[3], out_channels=out_channels[3], kernel_size=3, stride=2, padding=1
+                ),
+            ]
+        )
+        if config.readout_type == "project":
+            self.readout_projects = nn.ModuleList()
+            for _ in range(len(self.projects)):
+                self.readout_projects.append(
+                    nn.Sequential(nn.Linear(2 * config.hidden_size, config.hidden_size), ACT2FN[config.hidden_act])
+                )
+
```

Comment:
This code looks like is doing a lot of stuff. 

- It will break if the backbone returns more than 4 features, so you should create the `self.resize_layers` properly. From the paper, they fixed the target feature size to be the ones returned from the third stage of `ViT`. If we want this code to be general, `DPT` should stage a `target_size` and `later_hidden_states` and compute the rescaling factor based on them. 

A more elegant modeling design (IMHO) is to create a `DPTReassembleLayer`

```python
class DPTReassembleLayer(nn.Sequential):
    def __init__(self, target_size, in_size, in_channels, hidden_size, out_channels):
        super().__init__()
        should_upsample = in_size > target_size
        factor = in_size % target_size
        if should_upsample:
            self.resizer = nn.ConvTranspose2d(
                    in_channels=in_channels, out_channels=hidden_size, kernel_size=factor, stride=factor, padding=0
                )
        elif factor == 0:
            self.resizer = nn.Identity()
        elif not should_upsample:
            # so should downsample
            self.resizer = nn.Conv2d(
                    in_channels=in_channels, out_channels=hidden_size, kernel_size=3, stride=factor, padding=1
                )

        self.project = nn.Conv2d(in_channels=hidden_size, out_channels=out_channels, kernel_size=1)
```

Then you can create `n` layers based on the `backbone_hidden_sizes` and also polish up the forward method

Reply:
I've now taken this into account.